File: kernel/src/syscalls/brk.cpp
#include "process.h"

uintptr_t Process::do_sys_brk(uintptr_t addr) {
	dbgprintf("trying to set brk to %p, current is %p\n", addr, m_brk);
	if (addr < m_min_brk)
		return m_brk;

	uintptr_t next_page = PAGE_CEIL(m_brk);
	uintptr_t cur_page  = m_brk & PTL1_MASK;
	if (addr > next_page) {
		// We have to allocate space. Check wrapping first
		size_t sz = PAGE_CEIL(addr - next_page);
		if (next_page + sz < next_page)
			return m_brk;
		Range range(next_page, sz);
		if (!m_space.map_range(range, MemPerms::Read | MemPerms::Write))
			return m_brk;

	} else if (addr <= cur_page) {
		// Free space
		uintptr_t addr_next_page = PAGE_CEIL(addr);
		m_space.unmap_range({addr_next_page, next_page - addr_next_page});
	}

	dbgprintf("brk set to %p\n", addr);
	m_brk = addr;
	return m_brk;
}

File: kernel/src/process/syscalls/brk.zig
usingnamespace @import("../common.zig");
const mem = @import("../../mem/mem.zig");
const log = std.log.scoped(.sys_brk);

// sys_brk never returns an error. In case of an error, it just returns the
// current brk.

fn sys_brk(self: *Process, addr: usize) usize {
    log.debug("brk: trying to set to 0x{x}, current is 0x{x}\n", .{ addr, self.brk });
    if (addr < self.min_brk)
        return self.brk;

    const brk_next_page = mem.alignPageForward(self.brk);
    const brk_cur_page = mem.alignPageBackward(self.brk);
    const addr_next_page = mem.alignPageForwardChecked(addr) catch return self.brk;
    if (addr > brk_next_page) {
        // We have to allocate space
        const size = addr_next_page - brk_next_page;
        self.space.mapRange(brk_next_page, size, .{ .read = true, .write = true }, .{}) catch |err| switch (err) {
            error.OutOfMemory, error.NotUserRange => return self.brk,
            error.AlreadyMapped => unreachable,
        };
    } else if (addr <= brk_cur_page) {
        // We have to free space
        const size = brk_next_page - addr_next_page;
        self.space.unmapRange(addr_next_page, size) catch |err| switch (err) {
            error.NotMapped, error.NotUserRange => unreachable,
            // OOM should be impossible here, as we are not splitting any region.
            error.OutOfMemory => unreachable,
        };
    }

    log.debug("brk: set to 0x{x}\n", .{addr});
    self.brk = addr;
    return self.brk;
}

pub fn handle_sys_brk(self: *Process, arg0: usize) usize {
    const addr = arg0;
    return sys_brk(self, arg0);
}
